BFD-243: Update Contributing and README instructions (#292)
CMSgov/beneficiary-fhir-data: CONTRIBUTING.md

# Welcome!

We want to ensure a welcoming environment for all of our projects. Our staff follow the [18F Code of Conduct](https://github.com/18F/code-of-conduct/blob/master/code-of-conduct.md) and all contributors should do the same.

> Although this is a public repo, contributing to the BFD is for CMS approved contributors only, not outside contributors.

Reviewer: ```suggestion
Although this is a public repo, contributing to the BFD is for CMS-approved contributors only, not outside contributors.
```

Contributing to the BFD
---

### Background

BFD exists to enable the CMS Enterprise to drive innovation in data sharing so that beneficiaries and their healthcare partners have the data they need to make informed decisions about their healthcare.

> We provide a comprehensive, performant, and trustworthy platform to transform the way that the CMS enterprise shares and uses data by providing Coverage, Patient, and EOB FHIR-formatted data.

Reviewer: ```suggestion
We provide a comprehensive, performant, and trustworthy platform to transform the way that the CMS enterprise shares and uses data by providing [FHIR](https://www.hl7.org/fhir/)-formatted data, including beneficiary demographic, enrollment, and claims data.
```

> Review the [README](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/README.md) for additional information on the BFD. 

Reviewer: ```suggestion
Review the [README](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/README.md) for additional information on BFD. 
```

### Contributing changes

Contributions to the BFD are welcome from any party inside CMS.
Small changes like "good first issues" can be submitted for consideration without any additional process.

> Any [substantive change must go though an RFC process](#proposing-substantive-changes) before work on the change itself can start. 

Reviewer: ```suggestion
Any [substantive change must go though an RFC process](#proposing-substantive-changes) before work on the change itself can start.
```

> Any code changes should be properly commented and accompanied by appropriate unit test as per DASG engineering guidelines.

Reviewer 2: ```suggestion
Any code changes should be properly commented and accompanied by appropriate unit and integration test as per DASG engineering guidelines.
```

> Any code changes should be properly commented and accompanied by appropriate unit and integration test as per DASG engineering guidelines.

Reviewer: ```suggestion
Any code changes should be properly commented and accompanied by appropriate unit and integration test as per [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md).
```

> ![Alt text](docs/assets/BFDContributionDiagram.png "Diagram")

Reviewer: ```suggestion
![Workflow diagram for making a contribution to BFD](docs/assets/BFDContributionDiagram.png "Making a Contribution to BFD")
```

> ![Alt text](docs/assets/BFDContributionDiagram.png "Diagram")

Reviewer: @egkaufer Can you export this diagram as an SVG file, instead? That way, it'll be much easier for browsers and screen readers (for vision-impaired folks) to see.

**FAQ**

Q: What kind of changes don't require an RFC?

A: In general bug fixes and small changes that do not affect behavior or meaning. If you're unsure please quickly ask in the [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ) channel or in the Scrum of Scrums meeting. If your PR involves a substantial change, it will be rejected and you will be asked to go through the RFC process.

Q: How do I know what “first issues” are up for grabs? 

> A: First issues will be tracked on the BFD’s jira board and labeled (via versions) “Good First Issue”

Reviewer: ```suggestion
A: First issues are tracked on the BFD’s Jira board with the "Good_First_Issue" label: <https://jira.cms.gov/issues/?jql=project%20%3D%20BFD%20AND%20status%20%3D%20Open%20AND%20labels%20%3D%20Good_First_Issue>
```

### Proposing substantive changes

Substantive changes need to go through a design process involving the core team.
Opening an RFC provides a path for this inclusion and process.
Start an RFC by copying the file [`rfcs/0000-template.md`](rfcs/0000-template.md) to `rfcs/0000-<my idea>.md` and fill in the details. 
Open a PR using the RFC template [submit a pull request](#opening-a-pr).
The RFC will remain open for a 2 week period, at the end of which a go/no-go meeting will be held.

> If approved by at least two core team members and there are no outstanding reservations, the RFC will be accepted and given a number.

Reviewer 2: ```suggestion
If approved by at least two core team members and there are no outstanding reservations, the RFC will be accepted.
```

The RFC number is selected when issuing the PR, as the next available.

Q: What if I'm not sure if I need an RFC?

A: Reach out to the [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ) channel or ask in the Scrum of Scrums meeting and see what the BFD team thinks.

Q: How should I prepare for an RFC?

A: Bring the idea to the [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ) channel or the Scrum of Scrums meeting and talk it over with the core team.

Q: What if my RFC is not accepted?

A: It will be closed, but can be reopened if it is updated to address the items that prevented it's acceptance.

Q: What if my team doesn’t have the resources to implement our accepted RFC? 

A: Anyone can coordinate with you and the core team to take over the work. 

> ### Getting started

Reviewer: I'd suggest replacing this entire section with a brief pointer to one of these:

* <https://github.com/CMSgov/beneficiary-fhir-data/blob/master/README.md>
* <https://github.com/CMSgov/beneficiary-fhir-data/blob/master/dev/devenv-readme.md>

Reviewer 3: @cbrunefearless @keithdadkins @jzulim any preference?

Reviewer 4: The norm is to have a "Getting started" section in the main README that points to the CONTRIBUTING docs, or if it's simple enough, having the entire "getting started" instructions in the README (this is not likely what we want to do). 

If we want to maintain both local (no docker) and docker dev environments, we should have links to "Getting started with Docker" and "Getting started locally without Docker (Windows only)". The latter would point to the existing docs which is hard coded for Windows development until if/when we create directions for setting up dev environments on MacOS, Linux, etc.

Author: I concur with Keith on that..If the readme is simple enough, it should go in the readme, however more complex instructions should be outside of it.

Reviewer 2: I've simplified the 'Getting Started' section with pointers to Docker and Windows Local instructions.

# To make sure "origin" points to CMSgov/beneficiary-fhir-data run
# git remote -v
# if a different remote points to CMSgov/beneficiary-fhir-data
# replace "origin" with that remotes name in the origional command
```
In order to obtain permission to do this contact a github administrator or reach out on [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ).
Once pushed, [open a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) and use this [PR template](https://github.com/CMSgov/cms-oeda-dasg) found in .github

Please fill out each section of the body of the PR or quickly justify why the section does not apply to this particular change.
Reviewers will automatically be suggested or requested based on the projects CODEOWNERS file, feel free to add other reviewers if desired.
Once all automated checks pass and two reviewers have approved your pull request, a code owner from the core team will do a final review and make the decision to merge it in or not.

If you have any questions feel free to reach out on the [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ) channel in CMS slack!

> **FAQ**

Reviewer: Isn't this duplicative of the FAQ section above?

Reviewer 3: It looks like the question about a rejected PR (lines 121-123) was moved here - should we move it somewhere else?

Once pushed, [open a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) and use this [PR template](https://github.com/CMSgov/cms-oeda-dasg) found in .github

Please fill out each section of the body of the PR or quickly justify why the section does not apply to this particular change.
Reviewers will automatically be suggested or requested based on the projects CODEOWNERS file, feel free to add other reviewers if desired.
Once all automated checks pass and two reviewers have approved your pull request, a code owner from the core team will do a final review and make the decision to merge it in or not.

If you have any questions feel free to reach out on the [#bfd](https://cmsgov.slack.com/archives/C010WDXAZFZ) channel in CMS slack!

**FAQ**

Q: What if the core team rejects my PR?

A: The BFD core team will commit to never rejecting a PR without providing a path forward. The developer who put up the PR should review any feedback, and discuss with their product manager the scope of the work that is now outstanding.

> Setting up the FHIR server locally

Reviewer 2: I think we wanted to move this from the contributing and into the readme. Lines 125 through 288 need a new home, I believe.

Reviewer 3: I'm happy to move those to the README instead - @cbrunefearless @keithdadkins does that work?
